### ISSUE TITLE
[23000][1452] Cannot add or update a child row: a foreign key constraint fails (`sent_emails_url_clicked`, CONSTRAINT `sent_email_id` FOREIGN KEY (`id`) REFERENCES `sent_emails` (`id`) ON DELETE CASCADE ON UPDATE NO ACTION)

### ISSUE THREAD

**Issue by pcartagena74** [jdavidbakr/mail-tracker#39]
I've just started receiving this error using the package.

I sent 3 test emails with the test template.  With the third email, I added another link and it triggered the error above.

I've generated another email with links and none of them work either.   The click is registered b/c the email entry in the sent_emails table increments as expected.  

Any help would be appreciated and I'm happy to post whatever code snippets are necessary to figure this out.

**Comment by jdavidbakr** [comment]
Can you look in the error logs and see the actual query that's being made?  That means that for some reason the value being set for `sent_email_id` is not in the `sent_emails` table, since you see the table incrementing it makes me wonder what it's trying to insert.

**Comment by jdavidbakr** [comment]
Failing that, try a `dump($tracker)` in the `MailTrackerController::getL()` function inside the `if($tracker)` block.

**Comment by pcartagena74** [comment]
Thanks for a quick reply.

This is an example of a line of SQL as displayed in the Whoops output:
(SQL: insert into `sent_emails_url_clicked` (`sent_email_id`, `url`, `hash`, `updated_at`, `created_at`) values (4, http://www.msn.com, Llsy5wEvZswBlpjzJb1WRl6F5eRXDXYM, 2017-09-15 16:28:22, 2017-09-15 16:28:22))

I realize there are no quotes around the datetime fields and url but I believe that they're just not getting displayed here.

This is the output of the dump($tracker):
`SentEmail {#429 ▼
  #fillable: array:10 [▼
    0 => "hash"
    1 => "headers"
    2 => "sender"
    3 => "recipient"
    4 => "subject"
    5 => "content"
    6 => "opens"
    7 => "clicks"
    8 => "message_id"
    9 => "meta"
  ]
  #casts: array:1 [▼
    "meta" => "collection"
  ]
  #connection: "mysql"
  #table: null
  #primaryKey: "id"
  #keyType: "int"
  +incrementing: true
  #with: []
  #withCount: []
  #perPage: 15
  +exists: true
  +wasRecentlyCreated: false
  #attributes: array:15 [▼
    "id" => 4
    "campaignID" => 1
    "orgID" => 10
    "hash" => "Llsy5wEvZswBlpjzJb1WRl6F5eRXDXYM"
    "headers" => """
      Sender: PMI Office <office@pmimassbay.org>\r\n
      Message-ID: <39a1001002c6dcb51106c6512d425cdc@localhost>\r\n
      Date: Thu, 14 Sep 2017 23:35:32 -0400\r\n
      Subject: test email2\r\n
      From: PMI Office <office@pmimassbay.org>\r\n
      To: Phil Cartagena <pcartagena@partners.org>\r\n
      MIME-Version: 1.0\r\n
      Content-Type: text/html; charset=utf-8\r\n
      Content-Transfer-Encoding: quoted-printable\r\n
      X-Mailer-Hash: Llsy5wEvZswBlpjzJb1WRl6F5eRXDXYM\r\n
      """
    "sender" => "PMI Office <office@pmimassbay.org>"
    "recipient" => "Phil Cartagena <pcartagena@partners.org>"
    "subject" => "test email2"
    "content" => """
      <head>\n
      \tThis is the head\n
      </head>\n
      <body>\n
      \t<h1>\n
      \t\tThis is a test!\n
      \t</h1>\n
      \t<p>\n
      \t\tPlease\n
      \t\t<a href="http://www.google.com">Click Here</a>\n
      \t</p>\n
      \t<p>\n
      \t\t<a class="test" href="http://www.yahoo.com">\n
      \t\t\tClick here, too.\n
      \t\t</a>\n
      \t</p>\n
      \t<p>\n
      \t<a class="test" href="https://www.mcentric.org">\n
      \t\tBrought to you by https/mCentric\n
      \t</a>\n
      \t</p>\n
      \t<p>\n
      \t<a class="test" href="http://www.mcentric.org">\n
      \t\tBrought to you by http/mCentric\n
      \t</a>\n
      \t</p>\n
      \t<p>\n
      \t\t<a class="test" href="http://www.msn.com">\n
      \t\t\tBrought to you by http/msn\n
      \t\t</a>\n
      \t</p>\n
      \t<p>\n
      \t\t<a class="test" href="https://www.msn.com">\n
      \t\t\tBrought to you by https/msn\n
      \t\t</a>\n
      \t</p>\n
      </body>\n
      """
    "opens" => 5
    "clicks" => 9
    "message_id" => "39a1001002c6dcb51106c6512d425cdc@localhost"
    "meta" => "[]"
    "created_at" => "2017-09-14 23:35:32"
    "updated_at" => "2017-09-15 16:26:26"
  ]
  #original: array:15 [▼
    "id" => 4
    "campaignID" => 1
    "orgID" => 10
    "hash" => "Llsy5wEvZswBlpjzJb1WRl6F5eRXDXYM"
    "headers" => """
      Sender: PMI Office <office@pmimassbay.org>\r\n
      Message-ID: <39a1001002c6dcb51106c6512d425cdc@localhost>\r\n
      Date: Thu, 14 Sep 2017 23:35:32 -0400\r\n
      Subject: test email2\r\n
      From: PMI Office <office@pmimassbay.org>\r\n
      To: Phil Cartagena <pcartagena@partners.org>\r\n
      MIME-Version: 1.0\r\n
      Content-Type: text/html; charset=utf-8\r\n
      Content-Transfer-Encoding: quoted-printable\r\n
      X-Mailer-Hash: Llsy5wEvZswBlpjzJb1WRl6F5eRXDXYM\r\n
      """
    "sender" => "PMI Office <office@pmimassbay.org>"
    "recipient" => "Phil Cartagena <pcartagena@partners.org>"
    "subject" => "test email2"
    "content" => """
      <head>\n
      \tThis is the head\n
      </head>\n
      <body>\n
      \t<h1>\n
      \t\tThis is a test!\n
      \t</h1>\n
      \t<p>\n
      \t\tPlease\n
      \t\t<a href="http://www.google.com">Click Here</a>\n
      \t</p>\n
      \t<p>\n
      \t\t<a class="test" href="http://www.yahoo.com">\n
      \t\t\tClick here, too.\n
      \t\t</a>\n
      \t</p>\n
      \t<p>\n
      \t<a class="test" href="https://www.mcentric.org">\n
      \t\tBrought to you by https/mCentric\n
      \t</a>\n
      \t</p>\n
      \t<p>\n
      \t<a class="test" href="http://www.mcentric.org">\n
      \t\tBrought to you by http/mCentric\n
      \t</a>\n
      \t</p>\n
      \t<p>\n
      \t\t<a class="test" href="http://www.msn.com">\n
      \t\t\tBrought to you by http/msn\n
      \t\t</a>\n
      \t</p>\n
      \t<p>\n
      \t\t<a class="test" href="https://www.msn.com">\n
      \t\t\tBrought to you by https/msn\n
      \t\t</a>\n
      \t</p>\n
      </body>\n
      """
    "opens" => 5
    "clicks" => 9
    "message_id" => "39a1001002c6dcb51106c6512d425cdc@localhost"
    "meta" => "[]"
    "created_at" => "2017-09-14 23:35:32"
    "updated_at" => "2017-09-15 16:26:26"
  ]
  #dates: []
  #dateFormat: null
  #appends: []
  #events: []
  #observables: []
  #relations: []
  #touches: []
  +timestamps: true
  #hidden: []
  #visible: []
  #guarded: array:1 [▼
    0 => "*"
  ]
}`

**Comment by jdavidbakr** [comment]
Is there an entry is `sent_emails` for id `3`?

**Comment by pcartagena74** [comment]
Yes.  There are emails in the table for 1, 2, 3, 4.

**Comment by jdavidbakr** [comment]
I'm puzzled as to why you're getting that error, then.  What happens if you execute the query (properly quoted) directly into the database?

**Comment by pcartagena74** [comment]
It returns the same error about the foreign key constraint.  I'm puzzled as well.

**Comment by jdavidbakr** [comment]
Well, you could always drop that foreign key constraint manually, I'm not sure how I can further troubleshoot it without actually poking around in the database itself.

**Comment by pcartagena74** [comment]
I'm going to remove the package, clear the tables and re-fetch it with composer.  
The weirdest thing is that, 3 urls had worked (properly triggering the add to sent_emails_url_clicked table,  incrementing the clicks (when appropriate), and displaying the page) prior to encountering this error.

I introduced 1 change that I don't think should be the cause of this but figured I should mention:

I added a field to the sent_emails table (campaignID) that I plan to update when the event is triggered.  Right now it's just autopopulating with a 1. 

I'll post here after I've removed and replaced the package and tried again a few times.

**Comment by pcartagena74** [comment]
Ok, so I removed the package and emptied both tables.  Tested.  
I then re-added it and sent out a test message.

The sent_emails table has:
5	1	EUmANC49EfwPFKf6AkKnY6LUunuytXpj	"Sender: 
[long email with 5 links removed for brevity
"	2	0	82ef45d2305a478972be9f0e8ae4f56a@localhost	[]	2017-09-16 23:23:56	2017-09-16 23:25:51

Field 2 is the campaignID (of 1) that I added.

The error messages:
(1/2) PDOException
SQLSTATE[23000]: Integrity constraint violation: 1452 Cannot add or update a child row: a foreign key constraint fails (`bl5hod18fm0zc1zh`.`sent_emails_url_clicked`, CONSTRAINT `sent_email_id` FOREIGN KEY (`id`) REFERENCES `sent_emails` (`id`) ON DELETE CASCADE ON UPDATE NO ACTION)

(2/2) QueryException
SQLSTATE[23000]: Integrity constraint violation: 1452 Cannot add or update a child row: a foreign key constraint fails (`bl5hod18fm0zc1zh`.`sent_emails_url_clicked`, CONSTRAINT `sent_email_id` FOREIGN KEY (`id`) REFERENCES `sent_emails` (`id`) ON DELETE CASCADE ON UPDATE NO ACTION) (SQL: insert into `sent_emails_url_clicked` (`sent_email_id`, `url`, `hash`, `updated_at`, `created_at`) values (5, http://www.google.com, EUmANC49EfwPFKf6AkKnY6LUunuytXpj, 2017-09-16 23:26:26, 2017-09-16 23:26:26))

I deleted the foreign key constraint and it works fine.

I'm grateful that you've written this package and look forward to continuing to integrate it within my work.
